File: src/serve/instances.py
"""Discover and signal running serve instances by querying the OS.

No persistent state — `ps`/`lsof` are the source of truth.
"""

from __future__ import annotations

import os
import re
import signal
import subprocess
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Iterator


@dataclass
class Instance:
    pid: int
    port: int | None
    path: str
    mode: str
    started: str
    cmdline: str

    @property
    def url(self) -> str | None:
        return f"http://localhost:{self.port}" if self.port else None

    def to_dict(self) -> dict:
        d = asdict(self)
        d["url"] = self.url
        return d


_FLAGS_WITH_VALUE = {"-p", "--port", "--host"}
_BARE_FLAGS = {"--no-open", "--data-url"}


def _parse_serve_argv(argv: list[str]) -> str | None:
    """Given the full argv of a `serve` invocation (python + script + args),
    recover the file/dir argument the user passed (the first positional)."""
    # Skip the python interpreter and entry-point script (or `-m serve`).
    rest: list[str]
    if len(argv) >= 3 and argv[1] == "-m" and argv[2] == "serve":
        rest = argv[3:]
    elif len(argv) >= 2 and Path(argv[1]).name == "serve":
        rest = argv[2:]
    else:
        return None

    i = 0
    while i < len(rest):
        a = rest[i]
        if a in _FLAGS_WITH_VALUE:
            i += 2
            continue
        if a in _BARE_FLAGS or a.startswith("-"):
            i += 1
            continue
        return a
    return None


def _is_serve_invocation(argv: list[str]) -> bool:
    if len(argv) < 2:
        return False
    if Path(argv[1]).name == "serve":
        return True
    if argv[1] == "-m" and len(argv) >= 3 and argv[2] == "serve":
        return True
    return False


def _ps_processes() -> Iterator[tuple[int, str]]:
    """Yield (pid, cmdline) for every running process. Putting `command=` last
    in the format string lets `ps` extend the field without truncation."""
    result = subprocess.run(
        ["ps", "-axo", "pid=,command="],
        capture_output=True,
        text=True,
        check=True,
    )
    for line in result.stdout.splitlines():
        line = line.lstrip()
        if not line:
            continue
        pid_str, _, cmdline = line.partition(" ")
        try:
            pid = int(pid_str)
        except ValueError:
            continue
        yield pid, cmdline.strip()


_LSOF_PORT_RE = re.compile(r":(\d+)\s*\(LISTEN\)")


def _listening_port(pid: int) -> int | None:
    """Return the lowest TCP port `pid` is listening on, or None."""
    try:
        result = subprocess.run(
            ["lsof", "-a", "-nP", "-iTCP", "-sTCP:LISTEN", "-p", str(pid)],
            capture_output=True,
            text=True,
            check=False,
        )
    except FileNotFoundError:
        return None
    ports: list[int] = []
    for line in result.stdout.splitlines()[1:]:
        m = _LSOF_PORT_RE.search(line)
        if m:
            ports.append(int(m.group(1)))
    return min(ports) if ports else None


def _start_time(pid: int) -> str:
    """Return the human-readable start time of `pid`, or empty string."""
    try:
        result = subprocess.run(
            ["ps", "-o", "lstart=", "-p", str(pid)],
            capture_output=True,
            text=True,
            check=False,
        )
    except FileNotFoundError:
        return ""
    return result.stdout.strip()


def _process_cwd(pid: int) -> Path | None:
    """Return the working directory of `pid` via lsof, or None."""
    try:
        result = subprocess.run(
            ["lsof", "-a", "-d", "cwd", "-Fn", "-p", str(pid)],
            capture_output=True,
            text=True,
            check=False,
        )
    except FileNotFoundError:
        return None
    for line in result.stdout.splitlines():
        if line.startswith("n"):
            return Path(line[1:])
    return None


def _resolve_served_path(positional: str | None, cwd: Path | None) -> tuple[str, str]:
    """Replicate cli._resolve_path against the *target* process's cwd to
    recover the absolute path it's actually serving."""
    if positional is None:
        # serve with no arg: index.html in cwd if present, else cwd itself
        if cwd is None:
            return ("(default)", "?")
        index = cwd / "index.html"
        if index.is_file():
            return (str(index), "html")
        return (str(cwd), "directory")

    p = Path(positional)
    if not p.is_absolute() and cwd is not None:
        p = (cwd / p).resolve()
    elif p.is_absolute():
        p = p.resolve()

    if p.is_dir():
        return (str(p), "directory")
    suffix = p.suffix.lower()
    if suffix == ".md":
        return (str(p), "markdown")
    if suffix in {".html", ".htm"}:
        return (str(p), "html")
    return (str(p), "?")


def list_instances() -> list[Instance]:
    """Discover running serve instances. Excludes the current process."""
    self_pid = os.getpid()
    instances: list[Instance] = []
    for pid, cmdline in _ps_processes():
        if pid == self_pid:
            continue
        argv = cmdline.split()
        if not _is_serve_invocation(argv):
            continue
        # Skip subcommand invocations (serve list, serve kill, serve comments…).
        # Only the long-running document server has a listening port, but we
        # filter explicitly so transient subcommands don't briefly appear.
        rest_start = 3 if argv[1] == "-m" else 2
        rest = argv[rest_start:]
        if rest and rest[0] in {"list", "kill", "comments", "resolve", "agent-init", "ls"}:
            continue
        positional = _parse_serve_argv(argv)
        path, mode = _resolve_served_path(positional, _process_cwd(pid))
        port = _listening_port(pid)
        instances.append(
            Instance(
                pid=pid,
                port=port,
                path=path,
                mode=mode,
                started=_start_time(pid),
                cmdline=cmdline,
            )
        )
    instances.sort(key=lambda i: (i.port is None, i.port or 0, i.pid))
    return instances


def kill_instance(pid: int, force: bool = False) -> None:
    """Send SIGTERM (or SIGKILL with force=True) to `pid`.

    Raises ProcessLookupError if the process is gone, PermissionError if not
    permitted. Callers should translate these into user-facing errors.
    """
    sig = signal.SIGKILL if force else signal.SIGTERM
    os.kill(pid, sig)

File: instances.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"syscall"
)

// Instance represents a running serve process.
type Instance struct {
	PID     int    `json:"pid"`
	Port    int    `json:"port"`
	Path    string `json:"path"`
	Mode    string `json:"mode"`
	Started string `json:"started"`
	Cmdline string `json:"cmdline"`
}

func (i Instance) URL() string {
	if i.Port == 0 {
		return ""
	}
	return fmt.Sprintf("http://localhost:%d", i.Port)
}

var lsofPortRe = regexp.MustCompile(`:(\d+)\s*\(LISTEN\)`)

var subcommands = map[string]bool{
	"list": true, "ls": true, "kill": true,
	"comments": true, "resolve": true, "agent-init": true,
}

// listInstances discovers running serve-go instances via ps.
func listInstances() []Instance {
	selfPID := os.Getpid()
	out, err := exec.Command("ps", "-axo", "pid=,command=").Output()
	if err != nil {
		return nil
	}

	var instances []Instance
	for _, line := range strings.Split(string(out), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		pidStr, cmdline, found := strings.Cut(line, " ")
		if !found {
			continue
		}
		pid, err := strconv.Atoi(strings.TrimSpace(pidStr))
		if err != nil || pid == selfPID {
			continue
		}
		cmdline = strings.TrimSpace(cmdline)
		if !isServeInvocation(cmdline) {
			continue
		}
		argv := strings.Fields(cmdline)
		if isSubcommand(argv) {
			continue
		}
		positional := parseServeArgv(argv)
		cwd := processCWD(pid)
		path, mode := resolveServedPath(positional, cwd)
		port := listeningPort(pid)
		started := startTime(pid)
		instances = append(instances, Instance{
			PID:     pid,
			Port:    port,
			Path:    path,
			Mode:    mode,
			Started: started,
			Cmdline: cmdline,
		})
	}

	sort.Slice(instances, func(i, j int) bool {
		a, b := instances[i], instances[j]
		aPort, bPort := a.Port, b.Port
		if aPort == 0 {
			aPort = 1<<31 - 1
		}
		if bPort == 0 {
			bPort = 1<<31 - 1
		}
		if aPort != bPort {
			return aPort < bPort
		}
		return a.PID < b.PID
	})
	return instances
}

func isServeInvocation(cmdline string) bool {
	base := filepath.Base(strings.Fields(cmdline)[0])
	return base == "serve-go" || base == "serve"
}

func isSubcommand(argv []string) bool {
	if len(argv) < 2 {
		return false
	}
	return subcommands[argv[1]]
}

func parseServeArgv(argv []string) string {
	flagsWithValue := map[string]bool{"-p": true, "--port": true, "--host": true}
	i := 1
	for i < len(argv) {
		a := argv[i]
		if flagsWithValue[a] {
			i += 2
			continue
		}
		if strings.HasPrefix(a, "-") {
			i++
			continue
		}
		return a
	}
	return ""
}

func resolveServedPath(positional, cwd string) (string, string) {
	if positional == "" {
		if cwd == "" {
			return "(default)", "?"
		}
		index := filepath.Join(cwd, "index.html")
		if fi, err := os.Stat(index); err == nil && !fi.IsDir() {
			return index, "html"
		}
		return cwd, "directory"
	}
	p := positional
	if !filepath.IsAbs(p) && cwd != "" {
		p = filepath.Join(cwd, p)
	}
	p, _ = filepath.Abs(p)
	if fi, err := os.Stat(p); err == nil && fi.IsDir() {
		return p, "directory"
	}
	ext := strings.ToLower(filepath.Ext(p))
	switch ext {
	case ".md":
		return p, "markdown"
	case ".html", ".htm":
		return p, "html"
	default:
		return p, "?"
	}
}

func listeningPort(pid int) int {
	out, err := exec.Command("lsof", "-a", "-nP", "-iTCP", "-sTCP:LISTEN",
		"-p", strconv.Itoa(pid)).Output()
	if err != nil {
		return 0
	}
	minPort := 0
	for _, line := range strings.Split(string(out), "\n")[1:] {
		m := lsofPortRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		p, _ := strconv.Atoi(m[1])
		if minPort == 0 || p < minPort {
			minPort = p
		}
	}
	return minPort
}

func startTime(pid int) string {
	out, err := exec.Command("ps", "-o", "lstart=", "-p", strconv.Itoa(pid)).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func processCWD(pid int) string {
	out, err := exec.Command("lsof", "-a", "-d", "cwd", "-Fn",
		"-p", strconv.Itoa(pid)).Output()
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.HasPrefix(line, "n") {
			return line[1:]
		}
	}
	return ""
}

// killInstance sends SIGTERM (or SIGKILL) to pid.
func killInstance(pid int, force bool) error {
	sig := syscall.SIGTERM
	if force {
		sig = syscall.SIGKILL
	}
	proc, err := os.FindProcess(pid)
	if err != nil {
		return err
	}
	return proc.Signal(sig)
}
